test for bare expressions in test suite

smichr opened the pull request:
Bare expressions can represent a logical error that passes silently. e.g.
```
x + 1 instead of x += 1
x == 1 instead of x = 1
```
This PR adds checking for such expressions using ast.

It increases code-quality checking from 6 seconds to 32 seconds (based on observation of run-times of this and another PR).

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments

Got a tutorial on [ast here](https://www.mattlayman.com/blog/2018/decipher-python-ast/) and made this attempt at checking for relationals and binary ops that appear alone as an expression.

- [x] add tests -- where? I put code in the docstring which passes locally when run interactively; but this is not tested by the docstring tester, so it is there for show
- [x] get the body of if-blocks to be tested (and while-blocks, I suppose) -- see doctest with ">>"
    this turns out to not be necessary; by simply removing the If visitor handler, all Expr are traversed whereever they may appear. The one exception is code in a `with-block`; no checking is done in that context since the visitor handler for `with` does not process the nodes in that node.
- [x] only run on test files; code itself should be tested already

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
    * bare Relational (x == 1) and binary operations (x + 1) will raise a code-quality error if they are in the test suite and not in a `with-block`
<!-- END RELEASE NOTES -->

smichr committed:
closes #8357: test against bare expressions

Bare expressions can represent a logical error that passes silently. e.g.

x + 1 instead of x += 1
x == 1 instead of x = 1

smichr committed:
fix identified issues

smichr commented:
@oscarbenjamin said,

> This looks like a good check. I wonder if flake8 or similar can pick up on this.

Someone more familiar with that would have to see.

Interesting to see that there were many exceptions raised.

See XXX in the diff for questions. I am not sure what to do about the timing tests. For now I just skip files with "bench" in their name.

smichr commented on sympy/core/tests/test_arit.py:
```diff
@@ -1955,7 +1955,7 @@ def test_Mod():
 
     x_r, y_r = symbols('x_r y_r', real=True)
-    (Piecewise((x_r, y_r > x_r), (y_r, True)) / z) % 1
+    assert (Piecewise((x_r, y_r > x_r), (y_r, True)) / z) % 1
```
```suggestion
    assert unchanged(Mod, Piecewise((x_r, y_r > x_r), (y_r, True)) / z, 1)
```

smichr commented on sympy/simplify/tests/test_gammasimp.py:
```diff
@@ -67,8 +67,8 @@ def test_gammasimp():
     assert gammasimp(e**2/gamma(k + 1)) == k/gamma(k)
     a = R(1, 2) + R(1, 3)
     b = a + R(1, 3)
```
This was an assertion where the left and rhs of the `==` were disconnected; so searching for more than `==` expressions caught this.

oscarbenjamin commented on sympy/testing/tests/test_code_quality.py:
```diff
@@ -147,6 +196,11 @@ def test(fname):
             _test_this_file_encoding(fname, test_file)
 
     def test_this_file(fname, test_file):
+        idx = line_with_bare_expr(test_file.read()) if 'bench' not in fname else None  # XXX what should be done about bench tests
```
The benchmarks in the sympy repo are broken and should be removed so I think it's fine to skip them here.

oscarbenjamin commented on sympy/testing/tests/test_pytest.py:
```diff
@@ -43,7 +43,7 @@ def test_expected_exception_is_silent_with():
     try:
         with raises(Exception):
-            1 + 1
+            _ = 1 + 1
```
This kind of thing is less good but I don't know if there's a way to detect this.

smichr replied:
I'm not sure yet how to handle visitation of a "with" ( or "if") node. If the node includes the head and body of the block then I need to recurse into the body. Perhaps we can ignore contents of "with" blocks. But we need to look at the body of "if" blocks (per note in doctest).

smichr replied:
expressions in `with-blocks` are not tested

smichr committed:
fixing exposed tests

matrices/tests/test_matrices.py", line 1066, in test_simplify
polys/domains/tests/test_domains.py", line 989, in test_gaussian_domains
printing/tests/test_mathml.py", line 1252,
solvers/tests/test_solvers.py", line 175, in test_solve_args
utilities/tests/test_wester.py", line 1061, in test_M27
vector/tests/test_implicitregion.py", line 26, in test_regular_point
vector/tests/test_implicitregion.py", line 88, in test_rational_parametrization
geometry/tests/test_polygon.py", line 573, in test_first_moment

smichr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -1063,7 +1063,7 @@ def test_simplify():
     M.simplify()
     assert M == Matrix([[eq]])
-    M.simplify(ratio=oo) == M
+    assert M.simplify(ratio=oo) == M
```
```suggestion
    M.simplify(ratio=oo)
```

smichr commented on sympy/polys/domains/tests/test_domains.py:
```diff
@@ -983,10 +983,10 @@ def test_CC_double():
 def test_gaussian_domains():
     I = S.ImaginaryUnit
     a, b, c, d = [ZZ_I.convert(x) for x in (5, 2 + I, 3 - I, 5 - 5)]
```
```suggestion
    a, b, c, d = [ZZ_I.convert(x) for x in (5, 2 + I, 3 - I, 5 - 5*I)]
```

smichr commented on sympy/printing/tests/test_mathml.py:
```diff
@@ -1249,7 +1249,7 @@ class TestAssocOp(AssocOp):
     expr = TestAssocOp(1, 2)
-    mpp.doprint(expr) == \
+    assert mpp.doprint(expr) == \
         '<mrow><mi>testassocop</mi><mn>2</mn><mn>1</mn></mrow>'
```
```suggestion
        '<mrow><mi>testassocop</mi><mn>1</mn><mn>2</mn></mrow>'
```

smichr commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -172,7 +172,7 @@ def test_solve_args():
         (exp(x) - x, exp(y) - y)) == {x: -LambertW(-1), y: -LambertW(-1)}
     # --  when symbols given
-    solve([y, exp(x) + x], x, y) == [(-LambertW(1), 0)]
+    assert solve([y, exp(x) + x], x, y) == [(-LambertW(1), 0)]
```
```suggestion
    assert solve([y, exp(x) + x], x, y) == {y: 0, x: -LambertW(0)}
```

smichr commented on sympy/vector/tests/test_implicitregion.py:
```diff
@@ -85,7 +85,7 @@ def test_rational_parametrization():
 
     conic = ImplicitRegion((x, y), Eq(x**2 + 4*x*y + 3*y**2 + x - y + 10, 0))
-    conic.rational_parametrization(t) == (17/2 + 4/(3*t**2 + 4*t + 1), 4*t/(3*t**2 + 4*t + 1) - 11/2)
+    assert conic.rational_parametrization(t) == (17/2 + 4/(3*t**2 + 4*t + 1), 4*t/(3*t**2 + 4*t + 1) - 11/2)
```
```suggestion
    assert conic.rational_parametrization(t) == (S(17)/2 + 4/(3*t**2 + 4*t + 1), 4*t/(3*t**2 + 4*t + 1) - S(11)/2)
```

smichr commented on sympy/geometry/tests/test_polygon.py:
```diff
@@ -570,7 +570,7 @@ def test_first_moment():
 
     p2 = Polygon((0, 0), (12, 0), (12, 30))
-    p2.first_moment_of_area() == (1600/3, -640/3)
+    assert p2.first_moment_of_area() == (1600/3, -640/3)
```
```suggestion
    assert p2.first_moment_of_area() == (S(1600)/3, -S(640)/3)
```

smichr commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -1056,9 +1056,10 @@ def test_M26():
+    # TODO: Replace solve with solveset which gives both [+/- current answer]
+    # note that there is a typo in this test in the wester.pdf; there is no
+    # real solution for the equation as it appears in wester.pdf
+    assert solve(log(acos(asin(x**R(2, 3) - b)) - 1) + 2, x) == [b + sin(1 + cos(1/E**2 + 1))**R(3/2)]
```
```suggestion
    with assuming(b + sin(1 + cos(1/E**2 + 1))**R(3, 2) > 0):
        assert solve(log(acos(asin(x**R(2, 3) - b)) - 1) + 2, x) == [b + sin(1 + cos(1/E**2 + 1))**R(3, 2)]
```

smichr committed:
visitor only looks for Expr and With

smichr committed:
modify fixed tests

smichr commented:
If this tests pass, this is ready to go.

smichr commented:
The constraint of filtering only test files can be changed. At least we have a working check against changes that easily make their way into the test suite.
